File: sdks/ruby/test/node/typed_node_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "json"
# Require the pure-Ruby contract files directly (avoids loading the
# protobuf/grpc stack, so this runs without the gRPC gems installed).
require_relative "../../lib/blok/node/typed_node"

class TypedNodeTest < Minitest::Test
  class SearchNode < Blok::Node::TypedNode
    node_name "@acme/search"
    description "Full-text search"

    input do
      field :query, :string, required: true
      field :limit, :integer, default: 10
    end

    output do
      field :results, :array
      field :count, :integer
    end

    def run(_ctx, input)
      rows = [input[:query]] * input[:limit]
      { "results" => rows, "count" => rows.size }
    end
  end

  def test_validates_input_and_runs
    out = SearchNode.new.execute(nil, { "query" => "ada", "limit" => 2 })
    assert_equal 2, out["count"]
    assert_equal %w[ada ada], out["results"]
  end

  def test_applies_default_values
    out = SearchNode.new.execute(nil, { "query" => "x" })
    assert_equal 10, out["count"]
  end

  def test_missing_required_field_raises_structured_blok_error
    err = assert_raises(Blok::Errors::BlokError) do
      SearchNode.new.execute(nil, { "limit" => 3 })
    end
    assert_equal 400, err.http_status
    assert_equal "NODE_INPUT_VALIDATION", err.code
    assert_equal "@acme/search", err.node
  end

  def test_wrong_type_raises_structured_blok_error
    err = assert_raises(Blok::Errors::BlokError) do
      SearchNode.new.execute(nil, { "query" => 123 }) # Integer where :string declared
    end
    assert_equal 400, err.http_status
  end

  def test_reflection_schemas_and_description
    reflection = SearchNode.new.reflect
    assert_equal "Full-text search", reflection[:description]

    schema = JSON.parse(reflection[:input_schema_json])
    assert_equal "object", schema["type"]
    assert schema["properties"].key?("query")
    assert_equal ["query"], schema["required"]
    refute_nil reflection[:output_schema_json]
  end
end

File: sdks/php/tests/Node/TypedNodeTest.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Tests\Node;

use Blok\Blok\Errors\BlokError;
use Blok\Blok\Node\NodeReflector;
use Blok\Blok\Node\TypedNode;
use Blok\Blok\Types\Context;
use PHPUnit\Framework\TestCase;

final class SearchInput
{
    public function __construct(public string $query, public int $limit = 10)
    {
    }
}

final class SearchNode extends TypedNode
{
    public function name(): string
    {
        return '@acme/search';
    }

    public function description(): string
    {
        return 'Full-text search';
    }

    protected function inputClass(): string
    {
        return SearchInput::class;
    }

    protected function run(Context $ctx, object $input): mixed
    {
        /** @var SearchInput $input */
        $rows = array_fill(0, $input->limit, $input->query);

        return ['results' => $rows, 'count' => count($rows)];
    }
}

final class TypedNodeTest extends TestCase
{
    public function testValidatesInputAndRuns(): void
    {
        $out = (new SearchNode())->execute(new Context(), ['query' => 'ada', 'limit' => 2]);
        $this->assertSame(2, $out['count']);
        $this->assertSame(['ada', 'ada'], $out['results']);
    }

    public function testAppliesDefaultValues(): void
    {
        $out = (new SearchNode())->execute(new Context(), ['query' => 'x']);
        $this->assertSame(10, $out['count']);
    }

    public function testMissingRequiredFieldThrowsStructuredBlokError(): void
    {
        try {
            (new SearchNode())->execute(new Context(), ['limit' => 3]); // missing 'query'
            $this->fail('expected a BlokError');
        } catch (BlokError $e) {
            $this->assertSame(400, $e->httpStatus);
            $this->assertSame('NODE_INPUT_VALIDATION', $e->errorCode);
        }
    }

    public function testReflectsSchemasAndDescription(): void
    {
        $node = new SearchNode();
        $this->assertInstanceOf(NodeReflector::class, $node);
        $this->assertSame('Full-text search', $node->description());

        $schema = $node->inputSchema();
        $this->assertSame('object', $schema['type']);
        $props = (array) $schema['properties'];
        $this->assertArrayHasKey('query', $props);
        $this->assertSame(['query'], $schema['required']);
    }
}
